external/audiocodec/streaming/player.c

```diff
@@ -29,615 +29,555 @@
-
-            reachEOS = (totalBytesRead != bytesToRead);
-            remainingBytes += totalBytesRead;
-            tmp = buf;
-            continue;
-        }
-
-        if((tmp[0] != 0xff) || ((tmp[1] & 0xf6) != 0xf0))
-		{
-            ++pos;
-            ++tmp;
-            --remainingBytes;
-            continue;
-        }
-
-        //int protectionAbsent = tmp[1] & 0x01;
-        int frame_size = (tmp[3] & 0x03) << 11 | tmp[4] << 3 | tmp[5] >> 5;
-        //int headerSize = protectionAbsent ? 7 : 9;
-
-        // We found what looks like a valid frame,
-        // now find its successors.
-
-        ssize_t test_pos = pos + frame_size;
-
-        valid = true;
-        const int FRAME_MATCH_REQUIRED = 3;
-        for (int j = 0; j < FRAME_MATCH_REQUIRED; ++j)
-		{
-            uint8_t temp[6];
-            ssize_t retval = _source_read_at(fp, test_pos, temp, sizeof(temp));
-            if (retval < (ssize_t)sizeof(temp))
-			{
-                valid = false;
-                break;
-            }
-
-	        if((temp[0] != 0xff) || ((temp[1] & 0xf6) != 0xf0))
-			{
-	            valid = false;
-	            break;
-	        }
-
-	        int test_frame_size = (temp[3] & 0x03) << 11 | temp[4] << 3 | temp[5] >> 5;
-
-            test_pos += test_frame_size;
-        }
-
-        if (valid)
-            *inout_pos = pos;
-
-        ++pos;
-        ++tmp;
-        --remainingBytes;
-    } while (!valid);
-
-    return valid;
+	ssize_t pos = *inout_pos;
+	bool valid = false;
+
+	const int32_t kMaxReadBytes = 1024;
+	const int32_t kMaxBytesChecked = 4 * 1024;  //128 * 1024;
+	uint8_t buf[kMaxReadBytes];
+	ssize_t bytesToRead = kMaxReadBytes;
+	ssize_t totalBytesRead = 0;
+	ssize_t remainingBytes = 0;
+	bool reachEOS = false;
+	uint8_t *tmp = buf;
+
+	do {
+		if (pos >= *inout_pos + kMaxBytesChecked) {
+			break;              // Don't scan forever.
+		}
+
+		if (remainingBytes < 6) {
+			// syncword... frame length, 6 bytes in total
+			if (reachEOS) {
+				break;
+			}
+
+			memcpy(buf, tmp, remainingBytes);
+			bytesToRead = kMaxReadBytes - remainingBytes;
+
+			/*
+			 * The next read position should start from the end of
+			 * the last buffer, and thus should include the remaining
+			 * bytes in the buffer.
+			 */
+			totalBytesRead = _source_read_at(fp, pos + remainingBytes, buf + remainingBytes, bytesToRead);
+
+			if (totalBytesRead <= 0) {
+				break;
+			}
+
+			reachEOS = (totalBytesRead != bytesToRead);
+			remainingBytes += totalBytesRead;
+			tmp = buf;
+			continue;
+		}
+
+		if ((tmp[0] != 0xff) || ((tmp[1] & 0xf6) != 0xf0)) {
+			++pos;
+			++tmp;
+			--remainingBytes;
+			continue;
+		}
+		//int protectionAbsent = tmp[1] & 0x01;
+		int frame_size = (tmp[3] & 0x03) << 11 | tmp[4] << 3 | tmp[5] >> 5;
+		//int headerSize = protectionAbsent ? 7 : 9;
+
+		// We found what looks like a valid frame,
+		// now find its successors.
+
+		ssize_t test_pos = pos + frame_size;
+
+		valid = true;
+		const int FRAME_MATCH_REQUIRED = 3;
```

Comment:
This also can be defined.
#define MP3_FRAME_MATCH_REQUIRE 2
#define AAC_FRAME_MATCH_REQUIRE 3